File: apps/api/app/schemas/cs_goals_extraction.py
"""M15.1 — Candidate-goals extraction from VPD documents.

Mirrors the M16 MoM-extraction shape. The frontend takes the result,
presents a review modal with per-row checkboxes + inline edit, and on
confirm fans out POST /accounts/:id/cs-goals × N (followed by a PATCH
that attaches the suggested initiatives).

`extra="allow"` so the prompt can surface fields we haven't modelled
yet (e.g. confidence scores per goal) without a schema bump.
"""

from __future__ import annotations

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Mirrors apps/api/app/schemas/cs_goal.py CSGoalCategory + Initiative.stage.
CSGoalCategory = Literal[
    "cost_savings",
    "base_rationalization",
    "risk_mitigation",
    "adoption",
    "other",
]
InitiativeStage = Literal[
    "proposed", "committed", "in_flight", "implemented", "blocked", "cancelled"
]


class ExtractedInitiative(BaseModel):
    """One initiative under a goal. Maps 1:1 to cs_goal.Initiative."""

    model_config = ConfigDict(extra="allow")

    name: str
    description: str | None = Field(None, max_length=2000)
    stage: InitiativeStage | None = None
    levers: list[str] = Field(default_factory=list)


class ExtractedGoal(BaseModel):
    """One candidate goal. The first five fields map to CSGoalCreate;
    initiatives ship as a PATCH after create."""

    model_config = ConfigDict(extra="allow")

    title: str = Field(..., min_length=1, max_length=200)
    category: CSGoalCategory = "other"
    target_value: str | None = Field(None, max_length=200)
    target_date: date | None = None
    owner: str | None = Field(None, max_length=200)

    initiatives: list[ExtractedInitiative] = Field(default_factory=list)

    # Surfaced to the UI for the per-row tag (high/medium/low).
    confidence: Literal["high", "medium", "low"] | None = None
    rationale: str | None = Field(None, max_length=600)


class CsGoalsExtractionResult(BaseModel):
    model_config = ConfigDict(extra="allow")

    document_id: UUID | None = None
    goals: list[ExtractedGoal] = Field(default_factory=list)
    is_stub: bool = False

File: apps/web/src/types/cs_goals_extraction.ts
// Mirrors apps/api/app/schemas/cs_goals_extraction.py.

export type CSGoalCategory =
  | "cost_savings"
  | "base_rationalization"
  | "risk_mitigation"
  | "adoption"
  | "other";

export type InitiativeStage =
  | "proposed"
  | "committed"
  | "in_flight"
  | "implemented"
  | "blocked"
  | "cancelled";

export interface ExtractedInitiative {
  name: string;
  description?: string | null;
  stage?: InitiativeStage | null;
  levers: string[];
}

export interface ExtractedGoal {
  title: string;
  category: CSGoalCategory;
  target_value?: string | null;
  target_date?: string | null; // ISO YYYY-MM-DD
  owner?: string | null;
  initiatives: ExtractedInitiative[];
  confidence?: "high" | "medium" | "low" | null;
  rationale?: string | null;
}

export interface CsGoalsExtractionResult {
  document_id?: string | null;
  goals: ExtractedGoal[];
  is_stub: boolean;
}

export const CATEGORY_LABELS: Record<CSGoalCategory, string> = {
  cost_savings: "Cost savings",
  base_rationalization: "Base rationalization",
  risk_mitigation: "Risk mitigation",
  adoption: "Adoption",
  other: "Other",
};

export const CATEGORY_TONES: Record<CSGoalCategory, string> = {
  cost_savings: "bg-beroe-green/15 text-beroe-green border-beroe-green/30",
  base_rationalization: "bg-beroe-blue/10 text-beroe-blue border-beroe-blue/30",
  risk_mitigation: "bg-beroe-amber/15 text-beroe-amber border-beroe-amber/40",
  adoption: "bg-beroe-purple/10 text-beroe-purple border-beroe-purple/30",
  other: "bg-beroe-bg text-text-secondary border-beroe-card-border",
};

export const CATEGORIES: CSGoalCategory[] = [
  "cost_savings",
  "base_rationalization",
  "risk_mitigation",
  "adoption",
  "other",
];

export const CONFIDENCE_TONES: Record<"high" | "medium" | "low", string> = {
  high: "bg-beroe-green/15 text-beroe-green border-beroe-green/30",
  medium: "bg-beroe-amber/15 text-beroe-amber border-beroe-amber/40",
  low: "bg-beroe-bg text-text-secondary border-beroe-card-border",
};
